### PR TITLE
Compression Implementation

### PR DESCRIPTION
**Pull request by andradazoltan**
Documentation on the compression algorithm and the challenges it poses on the DPU implementation: https://docs.google.com/document/d/12mF_BAYqqlyH3QMWalj9xPZ6KcH32Unvpt-GKLbgdxU/edit

I also split out the decompression code to different files to make the main file easy to read and navigate. The decompression code does not need to be re-reviewed.

### CODE REVIEW

File: snappy/dpu-compress/dpu_compress.c [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+ * @param ptr: pointer to the value we want to hash
+ * @param shift: adjusts hash to be within table size
+ * @return Hash of four bytes stored at ptr
+ */
+static inline uint32_t hash(uint32_t ptr, int shift)
+{
+	uint32_t kmul = 0x1e35a7bd;
+	uint32_t bytes = read_uint32(ptr);
+	return (bytes * kmul) >> shift;
+}
```
Comment (andradazoltan) [inline snappy/dpu-compress/dpu_compress.c:158]: @craiig This is the hash function used by Snappy.

Comment (craiig) [reply snappy/dpu-compress/dpu_compress.c:158]: lmao, ok, pretty basic! I would definitely try the built-in hash function in that case. 

Here's the built in functions UpMem defined:
https://github.com/UBC-ECE-Sasha/upmem-sdk/blob/11b9a936c8ea79cdac856307f49b6c77d097eb54/dpu-rt/src/syslib/built_ins.h

Here's the docs on the hash function, it seems to be a simple xor
https://sdk.upmem.com/2019.4.0/210_DpuHandbook.html#hash

I would try to evaluate the new hash compared to this built in hash somehow - maybe try comparing compression ratios before and after to make sure they're about the same? What's the downside to a hash collision here anyways?

Comment (andradazoltan) [reply snappy/dpu-compress/dpu_compress.c:158]: The downside of a hash collision, is you replace a 4-byte match with something that doesn't fully match. When we are searching for these matches, it will be harder to find something which results in longer literals being stored and thus worse compression.

TLDR: more collisions = worse compression

Trying out their built hash function that results in the most bits (10 bits), I can definitely see a difference. For the largest file I can compress without multiple DPUs, I go from 36% to 25% compression.

However, it does let me reduce my hash table size from 32KB to 2KB which is quite helpful. Not sure what tradeoff I should be going for here...

Comment (andradazoltan) [reply snappy/dpu-compress/dpu_compress.c:158]: However...
It does reduce the cycle count by 20%.

Comment (jnider) [reply snappy/dpu-compress/dpu_compress.c:158]: Sometimes its possible to improve the distribution (i.e. reduce collisions) of simple hash functions by hashing twice with different algorithms/keys. Is that a possibility?

File: snappy/dpu-compress/dpu_compress.c [hunk 2]
```diff
@@ -35,17 +47,23 @@ static inline void advance_seqread(struct in_buffer_context *input, uint32_t len
...
-	mram_read(input_buf + offset, data_read, 16);
+	mram_read(&input_buf[WINDOW_ALIGN(offset, 8)], data_read, 16);
 
 	offset %= 8;
-	
 	return (data_read[offset] |
 			(data_read[offset + 1] << 8) |
 			(data_read[offset + 2] << 16) |
-			(data_read[offset + 3] << 24));
+			(data_read[offset + 3] << 24));
```
Comment (jnider) [inline snappy/dpu-compress/dpu_compress.c:56]: I really have nothing to say, so I'm going to pick on whitespace. It looks like the only change on this line is a space at the end of the line?

File: snappy/dpu-compress/dpu_compress.c (hunk 1 above)
Comment (craiig) [reply snappy/dpu-compress/dpu_compress.c:158]: I think I'm missing something - does their hash function only set 10 bits?

Comment (andradazoltan) [reply snappy/dpu-compress/dpu_compress.c:158]: Well it only XOR's 10 bits with each other. The remaining upper bits will be 0, if I understand this correctly. I did print out all the values when I was checking the function, and none are greater than 1024.

image

@jnider Yes, we can do that. We can really do whatever we want with the hash function.

File: snappy/dpu-compress/dpu_compress.c (hunk 2 above)
Comment (andradazoltan) [reply snappy/dpu-compress/dpu_compress.c:56]: I guess I added it by accident in my last commit...

**Comment by craiig** [comment]
Aha, I get it. I bet we can do something by combining the results of
multiple hash instructions together as Joel said. The manual appears to
tell us how to vary the precise bits that get xored using another operator.
I’ll poke at this a bit tomorrow!
